### PR TITLE
Inputs interface and fetch function

### LINKED ISSUE #137: Inputs core

### PR DESCRIPTION
**Pull request by mikouaj**
Fixes #137
Introduced input interface and fetch function

### CODE REVIEW

File: internal/inputs/inputs.go [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+	clusterID string
+	inputID   string
+	result    interface{}
+	err       error
+}
+
+//GetAllInputsData fetches data from given inputs for all given clusters in a concurrent manner
+func GetAllInputsData(inputs []Input, clusterIDs []string) (map[string]*Cluster, []error) {
+	tasksNo := len(inputs) * len(clusterIDs)
+	tasksChan := make(chan *getDataTask, tasksNo)
```
Comment (ewojtach) [inline internal/inputs/inputs.go:54]: This channel does not have to be that big. A size of `defaultMaxDataGetCoroutines` would be optimal if we fill it in separate goroutine.

File: internal/inputs/inputs.go [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+	resultsChan := make(chan *getDataTaskResult, tasksNo)
+	errorsChan := make(chan *getDataTaskResult, tasksNo)
+	log.Infof("creating %d get data tasks", tasksNo)
+
+	for _, input := range inputs {
```
Comment (ewojtach) [inline internal/inputs/inputs.go:59]: this can be filled in single goroutine

File: internal/inputs/inputs.go [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+		for _, clusterID := range clusterIDs {
+			tasksChan <- &getDataTask{input: input, clusterID: clusterID}
+		}
+	}
+	close(tasksChan)
+
+	log.Debugf("starting %d goroutines", defaultMaxDataGetCoroutines)
+	var wg sync.WaitGroup
```
Comment (ewojtach) [inline internal/inputs/inputs.go:67]: this processing can be done in one wrapping goroutine, to allow result and error channel to be significatly smaller (eq. `defaultMaxDataGetCoroutines`)
E.g.:
```
go func() {
    var wg sync.WaitGroup
   for i := 0; i < defaultMaxDataGetCoroutines; i++ {
		wg.Add(1)
		go getInputData(i, &wg, tasksChan, resultsChan, errorsChan)
	}
	wg.Wait()
	close(resultsChan)
	close(errorsChan)
}
```

File: internal/inputs/inputs.go [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+	close(resultsChan)
+	close(errorsChan)
+	log.Debugf("processing results and errors")
+	results := processResults(resultsChan)
+	errors := processErrors(errorsChan)
+	return results, errors
+}
+
+func getInputData(i int, wg *sync.WaitGroup, tasks chan *getDataTask, results chan *getDataTaskResult, errors chan *getDataTaskResult) {
+	defer wg.Done()
```
Comment (ewojtach) [inline internal/inputs/inputs.go:82]: why defer here instead of calling Done at end?

File: internal/inputs/inputs.go [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+	}
+	log.Debugf("goroutine %d done", i)
+}
+
+func processResults(resultsChan chan *getDataTaskResult) map[string]*Cluster {
+	results := make(map[string]*Cluster)
+	for result := range resultsChan {
+		data, ok := results[result.clusterID]
+		if !ok {
+			data = &Cluster{Data: make(map[string]interface{})}
```
Comment (ewojtach) [inline internal/inputs/inputs.go:102]: shouldn't we log the lack of data?

File: internal/inputs/inputs_test.go [hunk 6]
```diff
@@ -0,0 +1,179 @@
...
+		&inputMock{getIDFn: func() string { return "gke-api" }, getDataFn: func(clusterID string) (interface{}, error) { return "data", nil }},
+		&inputMock{getIDFn: func() string { return "kube-api" }, getDataFn: func(clusterID string) (interface{}, error) { return "data", nil }},
+		&inputMock{getIDFn: func() string { return "metrics-api" }, getDataFn: func(clusterID string) (interface{}, error) { return "data", nil }},
+		&inputMock{getIDFn: func() string { return "bad-api" }, getDataFn: func(clusterID string) (interface{}, error) { return nil, errors.New("error") }},
+	}
+	results, errors := GetAllInputsData(inputs, clusterIDs)
+	if len(results) != len(clusterIDs) {
+		t.Fatalf("number of results = %v; want %v", len(results), len(clusterIDs))
+	}
+	if len(errors) != len(clusterIDs) {
```
Comment (ewojtach) [inline internal/inputs/inputs_test.go:59]: shouldn't we expect empty error list?

Comment (mikouaj) [reply internal/inputs/inputs_test.go:59]: The test `inputs[]` contain one mock input ("bad-api") that produces an error on `getData()`.
Therefore, for each cluster, there will one error from this input. The idea was to cover error results processing as well.

File: internal/inputs/inputs.go (hunk 1 above)
Comment (mikouaj) [reply internal/inputs/inputs.go:54]: Good point, thank you!

File: internal/inputs/inputs.go (hunk 2 above)
Comment (mikouaj) [reply internal/inputs/inputs.go:59]: Good point, thank you!

File: internal/inputs/inputs.go (hunk 3 above)
Comment (mikouaj) [reply internal/inputs/inputs.go:67]: Good point, thank you!

File: internal/inputs/inputs.go (hunk 4 above)
Comment (mikouaj) [reply internal/inputs/inputs.go:82]: In this case the effect will be the same (function has no return statements).
However using defer in golang is a nice habit, as it always executes when the function returns. In this case it doesnt matter that much, but is useful i..e when the function has multiple return statements.

File: internal/inputs/inputs.go (hunk 5 above)
Comment (mikouaj) [reply internal/inputs/inputs.go:102]: It is not really lack of data, this is initialisation. This block will execute always for a cluster that is added to the results for the first time.

**Review by ewojtach — APPROVED** [review]

**Pull request merged**